File: mobile_apps/ios_app/TigerWallet/Models/BrandingConfig.swift
import Foundation
import SwiftUI

// MARK: - White-label branding

/// White-label branding config for the iOS app.
///
/// Loading order:
///  1. `WL_BRANDING_SLUG` Info.plist key (configurable per WL-client build).
///     Absent/empty => stock TigerWallet build; no remote fetch happens.
///  2. If a slug is set, fetch `GET {CONTROL_PLANE_URL}/api/v1/branding/{slug}`
///     on app startup (URLSession). The endpoint is PUBLIC (no auth) so a
///     WL-branded app needs no secrets.
///  3. Cache the fetched JSON in UserDefaults so a transient network failure
///     or cold start still shows the WL brand instead of TigerWallet.
///  4. Fall back to TigerWallet defaults if there is no slug, the fetch fails,
///     or the endpoint returns 404 (no WL client matches the slug).
///
/// `Info.plist` `CFBundleDisplayName` remains "TigerWallet" as the launcher
/// label default; [BrandingConfig.shared.appName] overrides the in-app
/// displayed name at runtime (the springboard label can't be changed at
/// runtime, so the in-app title + theme colors are what we override).
struct Branding: Codable, Equatable {
    var slug: String = ""
    var appName: String = "TigerWallet"
    var logoUrl: String = ""
    var primaryColor: String = "#FF6B35"
    var secondaryColor: String = "#1E3A5F"
    var domain: String = "tigerwallet.io"
    var supportEmail: String = "support@tigerwallet.io"
    var termsUrl: String = "https://tigerwallet.io/terms"
    var privacyUrl: String = "https://tigerwallet.io/privacy"

    /// TigerWallet stock branding — the backward-compatible default.
    static let defaults = Branding()

    /// Coding keys map snake_case JSON from the control plane to Swift fields.
    enum CodingKeys: String, CodingKey {
        case slug, appName = "app_name", logoUrl = "logo_url",
             primaryColor = "primary_color", secondaryColor = "secondary_color",
             domain, supportEmail = "support_email",
             termsUrl = "terms_url", privacyUrl = "privacy_url"
    }

    init(slug: String = "", appName: String = "TigerWallet", logoUrl: String = "",
         primaryColor: String = "#FF6B35", secondaryColor: String = "#1E3A5F",
         domain: String = "tigerwallet.io", supportEmail: String = "support@tigerwallet.io",
         termsUrl: String = "https://tigerwallet.io/terms",
         privacyUrl: String = "https://tigerwallet.io/privacy") {
        self.slug = slug
        self.appName = appName
        self.logoUrl = logoUrl
        self.primaryColor = primaryColor
        self.secondaryColor = secondaryColor
        self.domain = domain
        self.supportEmail = supportEmail
        self.termsUrl = termsUrl
        self.privacyUrl = privacyUrl
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        // Merge over defaults: missing/empty fields fall back to TigerWallet.
        func val(_ key: CodingKeys, _ dflt: String) -> String {
            let s = (try? c.decode(String.self, forKey: key)) ?? ""
            return s.isEmpty ? dflt : s
        }
        self.init(
            slug: (try? c.decode(String.self, forKey: .slug)) ?? "",
            appName: val(.appName, Branding.defaults.appName),
            logoUrl: val(.logoUrl, Branding.defaults.logoUrl),
            primaryColor: val(.primaryColor, Branding.defaults.primaryColor),
            secondaryColor: val(.secondaryColor, Branding.defaults.secondaryColor),
            domain: val(.domain, Branding.defaults.domain),
            supportEmail: val(.supportEmail, Branding.defaults.supportEmail),
            termsUrl: val(.termsUrl, Branding.defaults.termsUrl),
            privacyUrl: val(.privacyUrl, Branding.defaults.privacyUrl)
        )
    }
}

final class BrandingConfig: ObservableObject {
    static let shared = BrandingConfig()

    /// Published so SwiftUI views can react to async branding refreshes.
    @Published private(set) var branding: Branding

    private let defaults = UserDefaults.standard
    private let cacheKey = "wl_branding_json"
    private let slugKey = "wl_branding_slug"

    /// WL_BRANDING_SLUG from Info.plist (per WL-client build). Empty => stock.
    let slug: String

    /// CONTROL_PLANE_URL from Info.plist; falls back to local dev control plane.
    private let controlPlaneUrl: String

    private let session: URLSession

    private init() {
        let info = Bundle.main.infoDictionary ?? [:]
        self.slug = (info["WL_BRANDING_SLUG"] as? String ?? "").trimmingCharacters(in: .whitespaces)
        let cp = (info["WL_CONTROL_PLANE_URL"] as? String ?? "").trimmingCharacters(in: .whitespaces)
        self.controlPlaneUrl = cp.isEmpty ? "http://localhost:9008" : cp

        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        cfg.timeoutIntervalForResource = 20
        cfg.waitsForConnectivity = false
        self.session = URLSession(configuration: cfg)

        // Load cached branding synchronously so first frame is WL-branded
        // when a cache is present and matches the current build's slug.
        self.branding = BrandingConfig.loadCached(slug: self.slug, defaults: defaults)
    }

    /// Load from cache, then async-refresh from the control plane. Call from
    /// `AppDelegate.application(_:didFinishLaunchingWithOptions:)`.
    func bootstrap() {
        guard !slug.isEmpty else { return }
        refresh()
    }

    private static func loadCached(slug: String, defaults ud: UserDefaults) -> Branding {
        guard let data = ud.data(forKey: "wl_branding_json"),
              let decoded = try? JSONDecoder().decode(Branding.self, from: data) else {
            return Branding.defaults
        }
        // Only trust a cache that matches the current build's slug.
        guard decoded.slug == slug || (slug.isEmpty && decoded.slug.isEmpty) else {
            return Branding.defaults
        }
        return decoded
    }

    /// Fetch `GET /api/v1/branding/{slug}` and apply on success. Failures are
    /// silent — the cached/default branding remains (backward compatible).
    private func refresh() {
        let trimmed = slug.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? slug
        guard let url = URL(string: "\(controlPlaneUrl)/api/v1/branding/\(trimmed)") else { return }
        var req = URLRequest(url: url)
        req.httpMethod = "GET"

        let task = session.dataTask(with: req) { [weak self] data, response, _ in
            guard let self = self,
                  let http = response as? HTTPURLResponse,
                  (200..<300).contains(http.statusCode),
                  let data = data,
                  let decoded = try? JSONDecoder().decode(Branding.self, from: data) else {
                return
            }
            if let encoded = try? JSONEncoder().encode(decoded) {
                self.defaults.set(encoded, forKey: "wl_branding_json")
                self.defaults.set(self.slug, forKey: self.slugKey)
            }
            DispatchQueue.main.async {
                self.branding = decoded
            }
        }
        task.resume()
    }

    // --- Convenience accessors ---

    var appName: String { branding.appName }
    var logoUrl: String { branding.logoUrl }
    var primaryColor: String { branding.primaryColor }
    var secondaryColor: String { branding.secondaryColor }
    var domain: String { branding.domain }
    var supportEmail: String { branding.supportEmail }

    /// Parse the WL primary color into a SwiftUI Color (fails closed to default).
    var primarySwiftUIColor: Color { Color(hex: branding.primaryColor) ?? Color(hex: Branding.defaults.primaryColor)! }
    var secondarySwiftUIColor: Color { Color(hex: branding.secondaryColor) ?? Color(hex: Branding.defaults.secondaryColor)! }
}

// MARK: - Hex color helper

extension Color {
    /// Init from a `#RRGGBB` / `RRGGBB` / `#RRGGBBAA` hex string. nil on failure.
    init?(hex: String) {
        var s = hex.trimmingCharacters(in: .whitespaces)
        if s.hasPrefix("#") { s.removeFirst() }
        guard let v = UInt32(s, radix: 16) else { return nil }
        let r, g, b, a: Double
        switch s.count {
        case 6:
            r = Double((v >> 16) & 0xFF) / 255.0
            g = Double((v >> 8) & 0xFF) / 255.0
            b = Double(v & 0xFF) / 255.0
            a = 1.0
        case 8:
            r = Double((v >> 24) & 0xFF) / 255.0
            g = Double((v >> 16) & 0xFF) / 255.0
            b = Double((v >> 8) & 0xFF) / 255.0
            a = Double(v & 0xFF) / 255.0
        default:
            return nil
        }
        self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
    }
}

File: mobile_apps/flutter_app/lib/branding_config.dart
// TigerWallet Flutter — White-label branding config
//
// Loads WL branding for a rebranded Flutter build:
//   1. `--dart-define=WL_BRANDING_SLUG=xxx` (set per WL-client build).
//      Absent/empty => stock TigerWallet build; no remote fetch happens.
//   2. If a slug is set, fetch `GET {WL_CONTROL_PLANE_URL}/api/v1/branding/{slug}`
//      on app startup (via the `http` package, already a dep). The endpoint is
//      PUBLIC (no auth) so a WL-branded app needs no secrets.
//   3. Cache the fetched JSON in SharedPreferences so a transient network
//      failure / cold start still shows the WL brand instead of TigerWallet.
//   4. Fall back to TigerWallet defaults if there is no slug, the fetch fails,
//      or the endpoint returns 404 (no WL client matches the slug).
//
// BrandingConfig is a [ChangeNotifier]; ThemeProvider listens to it and
// rebuilds the Material theme with the WL primary/secondary colors whenever
// the branding refreshes.

import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

/// TigerWallet stock branding — the backward-compatible default.
class Branding {
  final String slug;
  final String appName;
  final String logoUrl;
  final String primaryColor;
  final String secondaryColor;
  final String domain;
  final String supportEmail;
  final String termsUrl;
  final String privacyUrl;

  const Branding({
    this.slug = '',
    this.appName = 'TigerWallet',
    this.logoUrl = '',
    this.primaryColor = '#FF6B35',
    this.secondaryColor = '#F7C94B',
    this.domain = 'tigerwallet.io',
    this.supportEmail = 'support@tigerwallet.io',
    this.termsUrl = 'https://tigerwallet.io/terms',
    this.privacyUrl = 'https://tigerwallet.io/privacy',
  });

  /// TigerWallet stock branding.
  static const Branding defaults = Branding();

  /// Merge a JSON map from the control plane over the TigerWallet defaults.
  /// Missing/empty fields fall back to defaults (never null/empty).
  factory Branding.fromJson(Map<String, dynamic> json, {String slug = ''}) {
    String pick(String key, String dflt) {
      final v = json[key];
      if (v is String && v.trim().isNotEmpty) return v.trim();
      return dflt;
    }

    return Branding(
      slug: slug,
      appName: pick('app_name', defaults.appName),
      logoUrl: pick('logo_url', defaults.logoUrl),
      primaryColor: pick('primary_color', defaults.primaryColor),
      secondaryColor: pick('secondary_color', defaults.secondaryColor),
      domain: pick('domain', defaults.domain),
      supportEmail: pick('support_email', defaults.supportEmail),
      termsUrl: pick('terms_url', defaults.termsUrl),
      privacyUrl: pick('privacy_url', defaults.privacyUrl),
    );
  }

  Map<String, dynamic> toJson() => {
        'slug': slug,
        'app_name': appName,
        'logo_url': logoUrl,
        'primary_color': primaryColor,
        'secondary_color': secondaryColor,
        'domain': domain,
        'support_email': supportEmail,
        'terms_url': termsUrl,
        'privacy_url': privacyUrl,
      };

  Branding copyWith({String? slug}) =>
      Branding(slug: slug ?? this.slug);

  @override
  String toString() => 'Branding(slug=$slug, appName=$appName, domain=$domain)';
}

/// Singleton white-label branding config. Notifies listeners on refresh.
class BrandingConfig extends ChangeNotifier {
  BrandingConfig._();
  static final BrandingConfig instance = BrandingConfig._();

  static const String _cacheKey = 'wl_branding_json';
  static const String _cacheSlugKey = 'wl_branding_slug';

  /// WL_BRANDING_SLUG from `--dart-define`. Empty => stock TigerWallet build.
  final String slug =
      const String.fromEnvironment('WL_BRANDING_SLUG', defaultValue: '').trim();

  /// Control plane base URL from `--dart-define`; defaults to local dev.
  final String controlPlaneUrl =
      const String.fromEnvironment('WL_CONTROL_PLANE_URL',
              defaultValue: 'http://localhost:9008')
          .trim();

  Branding _branding = Branding.defaults;
  bool _initialized = false;

  Branding get branding => _branding;

  // Convenience accessors.
  String get appName => _branding.appName;
  String get logoUrl => _branding.logoUrl;
  String get primaryColor => _branding.primaryColor;
  String get secondaryColor => _branding.secondaryColor;
  String get domain => _branding.domain;
  String get supportEmail => _branding.supportEmail;

  /// Parsed WL primary/secondary colors (falls closed to TigerWallet defaults).
  Color get primarySwatch =>
      parseColor(_branding.primaryColor) ??
      parseColor(Branding.defaults.primaryColor)!;
  Color get secondarySwatch =>
      parseColor(_branding.secondaryColor) ??
      parseColor(Branding.defaults.secondaryColor)!;

  /// Initialize from cache, then async-refresh from the control plane.
  /// Safe to call once at app startup; subsequent calls are no-ops.
  Future<void> init() async {
    if (_initialized) return;
    _initialized = true;

    await _loadCached();
    notifyListeners();

    if (slug.isNotEmpty) {
      // Fire-and-forget; refresh updates listeners on success.
      unawaited(refresh());
    }
  }

  Future<void> _loadCached() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final raw = prefs.getString(_cacheKey);
      if (raw == null) return;
      final decoded = jsonDecode(raw);
      if (decoded is Map<String, dynamic>) {
        final cached = Branding.fromJson(decoded, slug: slug);
        // Only trust a cache whose slug matches the current build's slug.
        if (slug.isEmpty || cached.slug == slug) {
          _branding = cached;
        }
      }
    } catch (_) {
      // Corrupt cache -> ignore, keep defaults.
    }
  }

  /// Fetch `GET {controlPlaneUrl}/api/v1/branding/{slug}` and apply on success.
  /// Failures (network, non-2xx, 404) are silent — cached/default branding
  /// remains (backward compatible).
  Future<void> refresh() async {
    if (slug.isEmpty) return;
    final uri = Uri.parse(
        '$controlPlaneUrl/api/v1/branding/${Uri.encodeComponent(slug)}');
    try {
      final res = await http
          .get(uri, headers: {'Accept': 'application/json'})
          .timeout(const Duration(seconds: 15));
      if (res.statusCode < 200 || res.statusCode >= 300) return; // 404 => no WL client
      final decoded = jsonDecode(res.body);
      if (decoded is Map<String, dynamic>) {
        _branding = Branding.fromJson(decoded, slug: slug);
        await _persist();
        notifyListeners();
      }
    } on SocketException {
      // Network failure — keep cached/default branding.
    } catch (_) {
      // Bad JSON / timeout — keep cached/default branding.
    }
  }

  Future<void> _persist() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_cacheKey, jsonEncode(_branding.toJson()));
      await prefs.setString(_cacheSlugKey, _branding.slug);
    } catch (_) {
      // Storage unavailable — branding stays in-memory.
    }
  }
}

/// Parse a `#RRGGBB` / `RRGGBB` / `#RRGGBBAA` hex string into a [Color].
/// Returns null on failure (callers fall back to TigerWallet defaults).
Color? parseColor(String hex) {
  var s = hex.trim();
  if (s.isEmpty) return null;
  if (s.startsWith('#')) s = s.substring(1);
  if (s.length == 6) s = 'FF$s';
  final v = int.tryParse(s, radix: 16);
  if (v == null) return null;
  return Color(v);
}
